api/python/t4/util.py

```diff
@@ -244,3 +244,42 @@ def get_local_registry():
 
 def get_remote_registry():
     return load_config().get('default_remote_registry', None)
+
+
+def quiltignore_filter(paths, ignore_rules, url_scheme):
+    """Given a list paths, filter out the paths which are captured by the given
+    ignore rules.
+
+    Args:
+        paths (list): a list or iterable of paths
+        ignore_rules (list): a list or iterable of ignore rules, in Unix shell
+            style wildcard format
+        url_scheme (str): the URL scheme, only the "file" scheme is currently
+            supported
+    """
+    if url_scheme == 'file':
+        from fnmatch import fnmatch
+
+        files, dirs = set(), set()
+        for path in paths:
+            if path.is_file():
+                files.add(path)
+            else:
+                dirs.add(path)
+
+        for ignore_rule in ignore_rules:
+            ignore_rule = os.getcwd() + '/' + ignore_rule
+
+            for dir in dirs:
+                # copy git behavior --- git matches paths and directories equivalently.
+                # e.g. both foo and foo/ will match the ignore rule "foo"
+                # but only foo/ will match the ignore rule "foo/"
+                if fnmatch(dir.as_posix() + "/", ignore_rule) or fnmatch(dir.as_posix(), ignore_rule):
+                    files = set(n for n in files if dir not in n.parents)
+                    dirs = dirs - {dir}
```

Comment:
Does this work for multiple dirs? I think it might break the iterator:
```
>>> foo = range(10)
>>> for x in foo:
...     foo.remove(x)
>>> foo
[1, 3, 5, 7, 9]
```